Add ability to select multiple needs for the Add / Edit Student Modal

namanhboi opened the pull request:
### Summary <!-- Required -->

I added a new React component with styling called Dropdownbox. It is a dropdown checklist of items, allowing for multiple selections, with a confirm button. I also used this component in the Add / Edit Student Modal to enable the selection of multiple needs. This is purely frontend and I haven't made this multiple selection work with the backend.

![image](https://github.com/cornell-dti/carriage-web/assets/61220757/9025a106-140b-4dc2-9307-8a125b0fe95a)
_When the box is not open_
![image](https://github.com/cornell-dti/carriage-web/assets/61220757/440690ae-6355-40a2-a725-3a2582b891e1)
_The box is opened and some items are selected_
![image](https://github.com/cornell-dti/carriage-web/assets/61220757/a56af219-1935-4a6f-a22e-be9834a20f16)
_The box is closed_


I still need to make the textbox for when "Other" is selected work with this new Dropdownbox component. My previous pull request (https://github.com/cornell-dti/carriage-web/pull/497) which implemented the textbox for "Other" was done under the assumption that the select react tag was used. I also need to discuss with Lazim about possible redesigns for the needs selection because I have a different opinion on where the textbox to enter an "Other" need should be placed.  


This pull request is a part of redesigning the Add / Edit Student Modal and specifically the needs selection.

- [x] Implemented multiple selection for needs in the Add / Edit Student Modal
- [x] Created a new React component for dropdown checklists.
- [ ] Need to make the textbox for when "Other" is selected work with this component. 


### Test Plan 

I tested the design on different display sizes and there seems to be an issue where the "Needs: " text isn't in bold while all the other ones are. 

![image](https://github.com/cornell-dti/carriage-web/assets/61220757/7cc4f305-4b8d-4254-8d3e-7de614cf7f49)
_On the Galaxy S20 inspect element_
![image](https://github.com/cornell-dti/carriage-web/assets/61220757/805a1077-aa69-420e-b48c-41762a4b1daf)
_On the Responsive size in inspect element_

Although none of the text are in bold when i'm just on the website and not in Inspect.

![image](https://github.com/cornell-dti/carriage-web/assets/61220757/aa27723b-296e-40bd-b8a2-5f213ef7e5c8)
_When I'm just on the website_

I will try to figure out this issue in the next PRs.

### Notes 

#### The "Other" needs textbox

In the future when I incorprate the textbox for when "Other" is selected, I need to discuss with Lazim about possible redesigns. This is because implementing the textbox inside the Dropdownbox component itself will make the component not as reusable. I think that the textbox should be moved out of the dropdown checklist and be placed below it.

![image](https://github.com/cornell-dti/carriage-web/assets/61220757/31fc6f48-93be-4443-9330-52d7a0836abd)
_Lazim's design_

![image](https://github.com/cornell-dti/carriage-web/assets/61220757/127edbca-4eea-4e3a-927b-79b7791d1a0b)
_My idea_


#### Styling of the component

I am not very good with css, so a lot of my styling is working directly with px values which might not be good.

#### Behavior when click outside of the dropdown box.

I haven't implemented this later on I will make it so that the box collapses. 

### Breaking Changes

I haven't found any so far.

Atikpui007 commented on frontend/src/components/DropdownBox/DropdownBox.tsx:
```diff
@@ -0,0 +1,81 @@
+
+      {isDropdownBoxOpen && (
+        <>
+          <div className="dropdown-list">
```
I noticed the dropdown box was custom styled and this is okay if there is a significant change from the native HTML styling but since they are very similar (essentially the same) lets stick to using native HTML elements as they are designed to be accessible and cross platform 

A suggestion is to map out everything and use `input type = checkbox`

Example 
```
      {isDropdownBoxOpen && (
        <div className="dropdown-list">
          {dataSource.map((itemName) => (
            <label key={itemName} className="dropdown-list-item">
              <input
                type="checkbox"
                checked={selectedItems.has(itemName)}
                onChange={() => toggleSelectItem(itemName)}
              />
              <span className="item-name">{itemName}</span>
            </label>
          ))}
```

Atikpui007 commented on frontend/src/components/Modal/RiderModalInfo.tsx:
```diff
@@ -131,22 +132,15 @@ const RiderModalInfo = ({
+          <DropdownBox
+            placeHolderText="Select Your Needs"
+            dataSource={Object.values(Accessibility)
+              .splice(1)
```
instead of using the splice we can just remove the None value from the Accessibility enum since the stakeholders suggested there is no need for it to be there

Atikpui007 reviewed (changes requested):
Overall a pretty good PR and hope you are enjoying working with Typescript. 
Remember to add the conditional render for the text input as well 
Note that it is rendered when `other` is selected so we can think of how to implement that in a generic way to support different uses of it

Atikpui007 commented on frontend/src/components/Modal/RiderModalInfo.tsx:
```diff
@@ -131,22 +132,15 @@ const RiderModalInfo = ({
-              );
-            })}
-          </select>
+          Needs:
```
The reason why this is not bolded when the screen size is changed is because it is not wrapped using a label. 

The headers for each input type uses a label and there is a media-query which keeps the label bolded on different screen sizes. Check out the code in `src/components/modal/ridermodal.module.css`

Needs should be wrapped like the other labels 
```
          <Label className={styles.label} htmlFor="needs">
            Needs:{' '}
          </Label>
```

namanhboi commented:
### Summary of changes: 

Added the option to include a text input box in the MultiselectBox component. You can specify the option that you want to be associated with the multiselectbox and when that option is selected, the text input box will appear. You can also specify the placeholder text for the text input box. 
![image](https://github.com/cornell-dti/carriage-web/assets/61220757/da791a2f-253f-4cd0-a2ef-8ba348a34106)


Some styling changes: 
- The MultiselectBox now uses default html checkbox styling instead of custom styling. 
- Wrapped the "Needs:" text around a Label, so now it is bold when the screen size changes.
- Changed px to rem in the css of the MultiselectBox component.

![image](https://github.com/cornell-dti/carriage-web/assets/61220757/b059bf31-910d-4bb9-a45f-39a0c3b7d219)
_This is what the header for the MultiselectBox looks like. Now, uses the correct arrow svg icons from figma_

![image](https://github.com/cornell-dti/carriage-web/assets/61220757/62213d49-f140-40ce-9073-6ed9f3065ba7)
_This is what it looks like when you open the MultiselectBox_

![image](https://github.com/cornell-dti/carriage-web/assets/61220757/6f71c205-7e6b-4ff1-898b-fd27bc64e595)
_This is what it looks like when you select some elements_

![image](https://github.com/cornell-dti/carriage-web/assets/61220757/5b4ee897-88f2-4cbe-8c2e-c08bf0dbf3d0)
_When you select the specified option, in this case "Other", the textbox will appear below with the placeholder text given_

![image](https://github.com/cornell-dti/carriage-web/assets/61220757/28011ea0-1e0d-4cdb-93ed-d918a3efa72e)
_The "Needs" text is now bold when you change screen sizes._

This pull request is a part of redesigning the Add / Edit Student Modal and specifically the needs selection.

- [x] Implemented multiple selection for needs in the Add / Edit Student Modal
- [x] Created a new React component for multiselect box.
- [x] Make the textbox for when "Other" is selected work with this component.

### Test plan

I have tested it in different screen sizes and found that it seems to scale fine on smaller screen sizes. This seems to be the case for the overall website. I will try to add media-query to the css file over the next few commits on the Add /Edit Student Modal.

![image](https://github.com/cornell-dti/carriage-web/assets/61220757/7472e321-6e98-42f6-a4c9-6653b476f32d)

![image](https://github.com/cornell-dti/carriage-web/assets/61220757/7f509560-98c5-4d17-bb90-d7361157105d)



### Notes: 
I haven't done the backend for this yet, but will start it soon. 

### Breaking changes

So far, I haven't noticed any yet, but I need to do more extensive testing. They will probably show up when I do the backend for the component.

Atikpui007 commented:
Closing because this requires lots of backend setup before frontend can be implemented
